AX: Serve AXLeftLineTextMarkerRangeForTextMarker, AXRightLineTextMarkerRangeForTextMarker, and AXStyleTextMarkerRangeForTextMarker off the main-thread

twilco opened the pull request:
#### c9870e67e54ff9ed67d6283ae72fcfd1b9e7b2bf
<pre>
AX: Serve AXLeftLineTextMarkerRangeForTextMarker, AXRightLineTextMarkerRangeForTextMarker, and AXStyleTextMarkerRangeForTextMarker off the main-thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=284564">https://bugs.webkit.org/show_bug.cgi?id=284564</a>
<a href="https://rdar.apple.com/141365741">rdar://141365741</a>

Reviewed by Chris Fleizach.

This commit adds accessibility-thread implementations for these attributes, and adds new layout tests to ensure we
behave correctly.

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/left-line-range-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/left-line-range.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/right-line-range-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/right-line-range.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/style-range-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/style-range.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::rangeWithSameStyle const):
(WebCore::AXTextMarker::lineRange const):
(WebCore::AXTextMarker::wordRange const):
(WebCore::AXTextMarker::sentenceRange const):
(WebCore::AXTextMarker::paragraphRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper lineTextMarkerRangeForTextMarker:forUnit:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::styleTextMarkerRangeForTextMarker):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::styleTextMarkerRangeForTextMarker):

Canonical link: <a href="https://commits.webkit.org/287990@main">https://commits.webkit.org/287990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/726d12d8961a9c79eea0fed4dce9e25af0088ded

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81525 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86141 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32519 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8869 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63671 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21363 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74200 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43960 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/659 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28369 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30972 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28962 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87493 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8758 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6216 "Found 1 new test failure: media/video-canvas-createPattern.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71963 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8939 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70017 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71196 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17729 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15248 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14166 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8718 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14246 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8556 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10363 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -633,6 +633,50 @@ AXTextMarker AXTextMarker::findLastBefore(std::optional<AXID> stopAtID) const
+        auto currentMarker = *this;
+
+        while (true) {
+            auto newMarker = currentMarker.findMarker(direction);
```
I imagine you considered this, but why would `findObjectWithRuns` not work here, if we only need to traverse objects? That would avoid the need to go marker-by-marker.

twilco replied:
That would only stop at objects that have runs, and for this API I'm not sure that's right. e.g. for this HTML:

```html
<div>
    <div>Foo</div>
    Bar
</div>

<div id="d1" style="font-style: italic">
    <div>Baz</div>
</div>
```

Given a text marker pointing to the first character of "Foo", should the style range end at the first spot of the "Baz"-static-text text marker? Or should it stop at the "anchor" position that actually has different style, which is `#d1`? I would guess the latter, though it probably doesn't really matter that much, especially in this contrived example. And we don't represent anchor positions at this point anyways 🤷

hoffmanjoshua replied:
I definitely see your point. Though, since we really only care about when text style differs, it seems safe to use the "equivalent"/deepest position that has runs for the perf benefit? I can see an argument either way.

twilco replied:
OK yeah, `findObjectWithRuns` is probably the most pragmatic choice right now — fixed in latest commit!

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -992,22 +1036,41 @@ bool AXTextMarker::isInTextRun() const
-    // FIXME: The other types aren't implemented yet.
-    RELEASE_ASSERT_NOT_REACHED();
+
+    ASSERT_NOT_REACHED();
```
We can probably remove this `ASSERT` now, since we have an `else` case as a catch-all.

twilco committed:
AX: Serve AXLeftLineTextMarkerRangeForTextMarker, AXRightLineTextMarkerRangeForTextMarker, and AXStyleTextMarkerRangeForTextMarker off the main-thread
https://bugs.webkit.org/show_bug.cgi?id=284564
rdar://141365741

Reviewed by Chris Fleizach.

This commit adds accessibility-thread implementations for these attributes, and adds new layout tests to ensure we
behave correctly.

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/left-line-range-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/left-line-range.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/right-line-range-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/right-line-range.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/style-range-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/style-range.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::rangeWithSameStyle const):
(WebCore::AXTextMarker::lineRange const):
(WebCore::AXTextMarker::wordRange const):
(WebCore::AXTextMarker::sentenceRange const):
(WebCore::AXTextMarker::paragraphRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper lineTextMarkerRangeForTextMarker:forUnit:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::styleTextMarkerRangeForTextMarker):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::styleTextMarkerRangeForTextMarker):

Canonical link: https://commits.webkit.org/287990@main
